### ISSUE TITLE
Warn of using back-references in a variable declaration statement

### ISSUE THREAD

**Issue by leagris** [koalaman/shellcheck#1653]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash
unset a b c
typeset -i a=2 b=$a c="$((a - 1))"
printf 'a=%d, b=%d, c=%d\n' $a $b $c
unset a b c
declare a=hello b=world c="$a $b"
printf 'a=%s, b=%s, c=%s\n' "$a" "$b" "$c"
```

Actual output:
```none
a=2, b=0, c=-1
a=hello, b=world, c= 
```

#### Here's what shellcheck currently says:

```none
$ shellcheck myscript
No issues detected!

$ 
```

#### Here's what I wanted or expected to see:

Back-referenced variables within the Bash's built-in declaration commands  `typeset`, `declare`, `local` or `readonly`, are not assigned their value; until the declaration command finishes processing its declared variables arguments.

If you need to back-reference a variable, to initialize another variable within a declaration command:

Assign the back_referee variable's value, before the declaration command back-referencing it.

##### Some problematic codes:

```Bash
# Result: referrer=
declare back_referee=referred_value referrer=$back_referee
```

```Bash
# SIGERR trap: bash: declare: `=42': not a valid identifier
declare back_referee='indirect_name' "$back_referee"=42
```

```Bash
# Result: SEC_DAY=0
readonly -i SEC_HOUR=3600 SEC_DAY=$((24 * SEC_HOUR))
```

```Bash
# Result: greeting="Hello !"
declare name="nobody" greeting="Hello ${name}!"
```

##### Safe code:

```Bash
declare back_referee=referred_value
declare referrer=$back_referee
```

##### Still valid code (*because it does not use Bash's built-in declaration commands*)
```Bash
back_referee=referred_value referrer=$back_referee
```

**Comment by koalaman** [comment]
This is now available as SC2318. Thanks!
